### ISSUE TITLE
docker-compose.yml needed (and added here)

### ISSUE THREAD

**Issue by rickcecil** [simon987/sist2#304]
Hi folks. I've been working off-and-on these past couple of days to get a working docker-compose file and I think I finally got one and wanted to share it here.

Step 1: Create the docker-compose.yml
```
version: '2'
services:

  sist2:
    image: simon987/sist2
    restart: unless-stopped
    volumes:
      - /path/to/data:/data
      - ./index:/index
    ports:
      - "4090:4090"
    command: web --bind 0.0.0.0:4090 --es-url http://es:9200 /data/<nameofindex>

  es:
    image: docker.elastic.co/elasticsearch/elasticsearch:7.14.0
    environment:
      - discovery.type=single-node
      - "ES_JAVA_OPTS=-Xms1G -Xmx2G"
```

Notes
- Make sure the internal port for sist2 matches the port for the external access included on the bind command.
- You can change the location of the index and data -- those are just my preferences
- I have not yet tested it, but it seems like, to add multiple indexes, you would just add any additional indexes to the end of the command, separated by spaces, like so:
- The sist2 webui will not work until the rest of the steps are complete.

```command: web --bind 0.0.0.0:4090 --es-url http://es:9200 /data/<nameofindex> /data/<nameofindex2> /data/<nameofindex3>```

Step 2: Scan your files
```docker-compose run sist2 scan /data/<nameofindex> -o /index/my_idx1```

Notes
- /data/archive is the mounted volume that you want to scan.
- my_idx1 is the name that will show up in sist2. Name it whatever you want.
- -o = output_folder
- Make sure that the output_folder is a mapped volume. 
- You can add -t # to add threading and make it go faster. But that is really down to the individual system.

```docker-compose run sist2 scan /data/archive -o /index/my_idx1 -t 4```

Step 3: Start the containers
```docker-compose up -d```

Step 4: Push the data to ElasticSearch
```docker-compose run sist2 index --force-reset --batch-size 1000 --es-url http://es:9200 /index/my_idx1```

If you get this error, you need to wait another minute and then try again.
```Getting error: [FATAL elastic.c] Could not get ES version ```

Note: Not sure if steps 3 and 4 should be switched, but this order worked for me...

Step 5: Start the containers, again.
```docker-compose up -d```

This time, you have the data imported into elasticsearch and the web ui for sist2 should start as expected.

**Comment by simon987** [comment]
Thanks @rickcecil 
I'll try to integrate it to the docs when I have time

**Comment by drewbitt** [comment]
```
 sist2:
    image: simon987/sist2
    restart: unless-stopped
    volumes:
      - /path/to/data:/data
      - ./index:/index
    ports:
      - "4090:4090"
    command: web --bind 0.0.0.0:4090 --es-url http://es:9200 /data/<nameofindex>
```

This container will not run on first start even if the directory exists. 

```
File not found: /data/nameofindex
[154F4DD0AA40] [2022-11-20 22:00:19] [FATAL serialize.c] Invalid/corrupt index (Could not find descriptor): /data/nameofindex/descriptor.json: No such file or directory
```

A scan must happen first to create an idx with a descriptor.json

**Comment by Nurech** [comment]
I made a compose for Windows to do all actions you need.

`start.bat`
```
set mypath=%cd%
@echo off
echo.
echo Active path: %mypath% 
echo.

rmdir /s /q .\my_index
mkdir documents

echo.
echo Made directory "documents". Move files there you want to be indexed.
echo.

docker pull docker.elastic.co/elasticsearch/elasticsearch:7.14.0 
docker pull simon987/sist2
docker-compose up -d
echo.
start chrome www.0.0.0.0:8080.com
pause
```
`docker-compose.yml`
```
---
version: "2.1" # For Windows users

services:

  # Starts elasticsearch
  elasticsearch:
    image: docker.elastic.co/elasticsearch/elasticsearch:7.14.0
    environment:
      - discovery.type=single-node
      - "ES_JAVA_OPTS=-Xms1G -Xmx2G"
    ports:
      - 9200:9200      
    healthcheck:
       test: curl -u elastic:elastic -s -f elasticsearch:9200/_cat/health >/dev/null || exit 1
       interval: 5s
       timeout: 60s
       retries: 60     

  # Scan the files and make a index
  sist2_scan:
    image: simon987/sist2
    container_name: sist2_scan
    restart: "no"
    volumes:
      - .\documents/:/tmp/es
      - .\my_index/:/my_index
    command: "scan --very-verbose --incremental ./my_index/idx -o ./my_index/idx /tmp/es"

  # Push index to elasticsearch
  sist2_index:
    image: simon987/sist2
    container_name: sist2_index
    restart: "no"
    depends_on:
      elasticsearch:
        condition: service_healthy    
    volumes:
      - .\documents/:/tmp/es
      - .\my_index/:/my_index
    command: "index --very-verbose --force-reset --batch-size 1000 --es-url http://elasticsearch:9200 ./my_index/idx"        
    
  # Start the web
  sist2_web:
    image: simon987/sist2
    container_name: sist2_web
    restart: "no"
    ports:
      - "8888:8888"    
    volumes:
      - .\documents/:/tmp/es
      - .\my_index/:/my_index
    command: "web --very-verbose --bind 0.0.0.0:8888 --es-url http://elasticsearch:9200 ./my_index/idx"    
    


volumes:
  documents:
    driver: local
  my_index:
    driver: local
```

**Comment by DOFfactory** [comment]
thanks for this docker-compose file. I tried to deploy it, the steps in elasticsearch and sist2_scan proceed in order, however, with the sist2_index I always get the error:

> `[FATAL elastic.c] Could not get ES version`

Do you have any suggestion how to fix this?

**Comment by DOFfactory** [comment]
OK, ignore my question, the issue was that they were not in the same subnet. Even the containers `sist2_scan` and `sist2_index` need to be in the same subnet, otherwise it is not working.

**Comment by Nurech** [comment]
Here's a Linux version:
Default driver is bridge, so there shouldn't be need to define a network, but I added one anyway.
Web should give you an indexed test file after everything is running: http://0.0.0.0:9200/idx/_search

`start.sh`
```
#!/bin/bash
mkdir documents
mkdir my_index
touch documents/test.txt
echo 'This is a test content' -> documents/test.txt
docker kill $(docker ps -q)
docker compose up -d
```
`docker-compose.yml`
```
version: '2.5'

services:

  # 1 Start elasticsearch
  elasticsearch:
    image: docker.elastic.co/elasticsearch/elasticsearch:7.14.0
    environment:
      - discovery.type=single-node
      - "ES_JAVA_OPTS=-Xms1G -Xmx4G"
    ports:
      - "9200:9200"
    expose:
    - "9200"
    networks:
      - sist2      
    healthcheck:
      test: curl -u elastic:elastic -s -f elasticsearch:9200/_cat/health >/dev/null || exit 1
      interval: 30s
      timeout: 10s
      retries: 5

  # 2 Scan the files and make a index
  sist2_scan:
    image: simon987/sist2
    container_name: sist2_scan
    restart: "no"
    depends_on:
      elasticsearch:
        condition: service_healthy
    networks:
      - sist2      
    volumes:
      - ./documents/:/tmp/es
      - ./my_index/:/my_index
    command: "scan --very-verbose --incremental ./my_index/idx -o ./my_index/idx /tmp/es/"

  # 3 Push index to elasticsearch
  sist2_index:
    image: simon987/sist2
    container_name: sist2_index
    restart: "no"
    networks:
      - sist2  
    depends_on:
      - elasticsearch
      - sist2_scan
    volumes:
      - ./documents/:/tmp/es
      - ./my_index/:/my_index
    command: "index --very-verbose --es-index idx --force-reset --batch-size 1000 --es-url http://elasticsearch:9200 ./my_index/idx"
    
  # 4 Start the web UI
  sist2_web:
    image: simon987/sist2
    container_name: sist2_web
    restart: "no"
    networks:
      - sist2
    depends_on:
      - elasticsearch
      - sist2_index
    ports:
      - "8888:8888"
    volumes:
      - ./documents/:/tmp/es
      - ./my_index/:/my_index
    command: "web --very-verbose --es-index idx --bind 0.0.0.0:8888 --es-url http://0.0.0.0:9200 ./my_index/idx"

volumes:
  documents:
    driver: local
  my_index:
    driver: local

networks:
  sist2:   
 ```

**Comment by simon987** [comment]
Closing this in favor of the compose file in the README.md
